sdks/java/core/src/main/java/org/apache/beam/sdk/io/DefaultFilenamePolicy.java

```diff
@@ -0,0 +1,188 @@
+import org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.options.ValueProvider.NestedValueProvider;
+import org.apache.beam.sdk.options.ValueProvider.StaticValueProvider;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+
+/**
+ * A default {@link FilenamePolicy} for unwindowed files. This policy takes four parameters:
+ *
+ * <ul>
+ *   <li>{@code baseOutputDirectory}, a {@link ResourceId} representing where the output files will
+ *        be written.
+ *   <li>{@code prefix}, {@code shardTemplate}, and {@code suffix}, which will together will be
+ *        passed to {@link #constructName} to form the name of files written to the the output
+ *        directory.
+ *   <li>For more information on the output filename, see {@link #constructName}.
+ * </ul>
+ *
+ * <p>Most users of unwindowed files will use this {@link DefaultFilenamePolicy}. For more advanced
+ * uses in generating different files for each window and other sharding controls, see the
+ * {@code WriteOneFilePerWindow} example pipeline.
+ */
+public final class DefaultFilenamePolicy extends FilenamePolicy {
+  /** The default sharding name template used in {@link #constructUsingStandardParameters}. */
+  public static final String DEFAULT_SHARD_TEMPLATE = ShardNameTemplate.INDEX_OF_MAX;
+
+  // Pattern that matches shard placeholders within a shard template.
+  private static final Pattern SHARD_FORMAT_RE = Pattern.compile("(S+|N+)");
+
+  /**
+   * Constructs a new {@link DefaultFilenamePolicy}.
+   *
+   * @see DefaultFilenamePolicy for more information on the arguments to this function.
+   */
+  public DefaultFilenamePolicy(String prefix, String shardTemplate, String suffix) {
+    this(StaticValueProvider.of(prefix), shardTemplate, suffix);
+  }
+
+  /**
+   * Constructs a new {@link DefaultFilenamePolicy}.
+   *
+   * @see DefaultFilenamePolicy for more information on the arguments to this function.
+   */
+  public DefaultFilenamePolicy(ValueProvider<String> prefix, String shardTemplate, String suffix) {
+    this.prefix = prefix;
+    this.shardTemplate = shardTemplate;
+    this.suffix = suffix;
+  }
+
+  /**
+   * A helper function to construct a {@link DefaultFilenamePolicy} using the standard filename
+   * parameters, namely a provided {@link ResourceId} for the output prefix, and possibly-null
+   * shard name template and suffix.
+   *
+   * <p>Any filename component of the provided resource will be used as the filename prefix.
+   *
+   * <p>If provided, the shard name template will be used; otherwise {@link #DEFAULT_SHARD_TEMPLATE}
+   * will be used.
+   *
+   * <p>If provided, the suffix will be used; otherwise the files will have an empty suffix.
+   */
+  public static DefaultFilenamePolicy constructUsingStandardParameters(
```

Comment:
A bit odd to have public constructors AND a public factory function. Can you make some of them package-local, or at least document the target audience of each?